### PR TITLE
Wrap getCompiledKernel's return value into a struct. 

### LINKED ISSUE #775: Explore changing the eviction policy for broadcasted tensors that are loaded from global memory
For all loaded tensors, we currently stream in the Tensors with an evict first policy as is represented by the `.cop` field value of `.cs`.

https://github.com/NVIDIA/Fuser/blob/7aaa6c3487e29c40ec91c42b58019dd289654568/runtime/array.cu#L205C1-L208C43

```python
        uint4& data = *reinterpret_cast<uint4*>(to);
        asm volatile("ld.global.cs.v4.s32 {%0,%1,%2,%3}, [%4];"
                     : "=r"(data.x), "=r"(data.y), "=r"(data.z), "=r"(data.w)
                     : "l"((uint4*)from));
```

There are notable circumstances, particularly when we pass in broadcasted masks in dot-product attention, where it is beneficial to cache those tensors in L2 as they get reused across blocks.  We are currently hurting are perf by up to 10% by not caching the broadcasted masks.  It is likely we don't want to liberally set the the cache policy but add an option on our `loadGlobalToLocal` intrinsic to use a different policy.  The beneficial policy for cache in L2 is `.cg`.

Here are a couple repros where it matters:
The forward pass of causual dot-product attention.  The kernel time goes from `1.014ms` -> `0.901ms` on A100 with the cache policy changed.  NOTE: the arrow points to the tensor that likely matters.

```python
import torch
from nvfuser import FusionDefinition, DataType

inputs = [
    torch.randn(16, 25, 512, 512, device="cuda"),
    torch.randn(1, 1, 512, 512, device="cuda"),
]

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=DataType.Float, is_cpu=False)
--->T1 = fd.define_tensor(shape=[1, 1, -1, -1], contiguity=[None, None, False, True], dtype=DataType.Float, is_cpu=False)
    S2 = fd.define_scalar(0.125000, dtype=DataType.Float)
    T3 = fd.ops.mul(T0, S2)
    S4 = fd.define_scalar(0.00000, dtype=DataType.Float)
    T5 = fd.ops.eq(T1, S4)
    S6 = fd.define_scalar(16, dtype=DataType.Int)
    S7 = fd.define_scalar(25, dtype=DataType.Int)
    S8 = fd.define_scalar(512, dtype=DataType.Int)
    S9 = fd.define_scalar(512, dtype=DataType.Int)
    V10 = fd.define_vector([S6, S7, S8, S9], dtype=DataType.Int)
    T11 = fd.ops.broadcast_in_dim(T5, shape=V10, broadcast_dims=[0, 1, 2, 3])
    S12 = fd.define_scalar(float("-inf"), dtype=DataType.Float)
    T13 = fd.ops.where(T11, S12, T3)
    T14 = fd.ops.max(T13, axes=[3], keepdim=False, dtype=DataType.Null)
    S15 = fd.define_scalar(16, dtype=DataType.Int)
    S16 = fd.define_scalar(25, dtype=DataType.Int)
    S17 = fd.define_scalar(512, dtype=DataType.Int)
    S18 = fd.define_scalar(1, dtype=DataType.Int)
    V19 = fd.define_vector([S15, S16, S17, S18], dtype=DataType.Int)
    T20 = fd.ops.broadcast_in_dim(T14, shape=V19, broadcast_dims=[0, 1, 2])
    S21 = fd.define_scalar(16, dtype=DataType.Int)
    S22 = fd.define_scalar(25, dtype=DataType.Int)
    S23 = fd.define_scalar(512, dtype=DataType.Int)
    S24 = fd.define_scalar(512, dtype=DataType.Int)
    V25 = fd.define_vector([S21, S22, S23, S24], dtype=DataType.Int)
    T26 = fd.ops.broadcast_in_dim(T20, shape=V25, broadcast_dims=[0, 1, 2, 3])
    T27 = fd.ops.sub(T13, T26)
    T28 = fd.ops.exp(T27)
    T29 = fd.ops.sum(T28, axes=[3], keepdim=False, dtype=DataType.Null)
    S30 = fd.define_scalar(16, dtype=DataType.Int)
    S31 = fd.define_scalar(25, dtype=DataType.Int)
    S32 = fd.define_scalar(512, dtype=DataType.Int)
    S33 = fd.define_scalar(1, dtype=DataType.Int)
    V34 = fd.define_vector([S30, S31, S32, S33], dtype=DataType.Int)
    T35 = fd.ops.broadcast_in_dim(T29, shape=V34, broadcast_dims=[0, 1, 2])
    S36 = fd.define_scalar(16, dtype=DataType.Int)
    S37 = fd.define_scalar(25, dtype=DataType.Int)
    S38 = fd.define_scalar(512, dtype=DataType.Int)
    S39 = fd.define_scalar(512, dtype=DataType.Int)
    V40 = fd.define_vector([S36, S37, S38, S39], dtype=DataType.Int)
    T41 = fd.ops.broadcast_in_dim(T35, shape=V40, broadcast_dims=[0, 1, 2, 3])
    T42 = fd.ops.reciprocal(T41)
    T43 = fd.ops.mul(T28, T42)
    S44 = fd.define_scalar(0.00000, dtype=DataType.Float)
    S45 = fd.define_scalar(1.00000, dtype=DataType.Float)
    S46 = fd.define_scalar(16, dtype=DataType.Int)
    S47 = fd.define_scalar(25, dtype=DataType.Int)
    S48 = fd.define_scalar(512, dtype=DataType.Int)
    S49 = fd.define_scalar(512, dtype=DataType.Int)
    T50 = fd.ops.uniform(S44, S45, shape=[S46, S47, S48, S49], dtype=DataType.Float)
    S51 = fd.define_scalar(0.900000, dtype=DataType.Float)
    T52 = fd.ops.lt(T50, S51)
    T53 = fd.ops.cast(T52, dtype=DataType.Float)
    T54 = fd.ops.mul(T43, T53)
    S55 = fd.define_scalar(1.11111, dtype=DataType.Float)
    T56 = fd.ops.mul(T54, S55)
    fd.add_output(T11)
    fd.add_output(T43)
    fd.add_output(T52)
    fd.add_output(T56)

with FusionDefinition() as fd:
    nvfuser_fusion_id1(fd)

for _ in range(5):
    out = fd.execute(inputs)
```

Here is the backward pass where the time goes from `850us` -> `798us`.

```python
import torch
from nvfuser import FusionDefinition, DataType

inputs = [
    1.1111111,
    1.0,
    1,
    1,
    1,
    torch.randn(1, 1, 512, 512, device="cuda") > 0.0,
    torch.randn(16, 25, 512, 512, device="cuda"),
    torch.randn(16, 25, 512, 512, device="cuda") > 0.0,
    torch.randn(16, 25, 512, 512, device="cuda"),
]

def nvfuser_fusion_id9(fd : FusionDefinition) -> None :
    S0 = fd.define_scalar(None, dtype=DataType.Double)
    S1 = fd.define_scalar(None, dtype=DataType.Double)
    S2 = fd.define_scalar(None, dtype=DataType.Int)
    S3 = fd.define_scalar(None, dtype=DataType.Int)
    S4 = fd.define_scalar(None, dtype=DataType.Int)
    T5 = fd.define_tensor(shape=[1, 1, -1, -1], contiguity=[None, None, True, True], dtype=DataType.Bool, is_cpu=False)
    T6 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T7 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=DataType.Bool, is_cpu=False)
    T8 = fd.define_tensor(shape=[-1, -1, -1, -1], contiguity=[True, True, True, True], dtype=DataType.Float, is_cpu=False)
    T9 = fd.ops.cast(T7, dtype=DataType.Float)
    T10 = fd.ops.mul(T8, S1)
    T11 = fd.ops.mul(T10, T9)
    T12 = fd.ops.mul(T6, T11)
    T13 = fd.ops.sum(T12, axes=[3], keepdim=False, dtype=DataType.Null)
    S14 = fd.define_scalar(16, dtype=DataType.Int)
    S15 = fd.define_scalar(25, dtype=DataType.Int)
    S16 = fd.define_scalar(512, dtype=DataType.Int)
    S17 = fd.define_scalar(1, dtype=DataType.Int)
    V18 = fd.define_vector([S14, S15, S16, S17], dtype=DataType.Int)
    T19 = fd.ops.broadcast_in_dim(T13, shape=V18, broadcast_dims=[0, 1, 2])
    S20 = fd.define_scalar(16, dtype=DataType.Int)
    S21 = fd.define_scalar(25, dtype=DataType.Int)
    S22 = fd.define_scalar(512, dtype=DataType.Int)
    S23 = fd.define_scalar(512, dtype=DataType.Int)
    V24 = fd.define_vector([S20, S21, S22, S23], dtype=DataType.Int)
    T25 = fd.ops.broadcast_in_dim(T19, shape=V24, broadcast_dims=[0, 1, 2, 3])
    T26 = fd.ops.sub(T11, T25)
    T27 = fd.ops.mul(T6, T26)
    S28 = fd.define_scalar(0.00000, dtype=DataType.Double)
    T29 = fd.ops.where(T5, S28, T27)
    T30 = fd.ops.mul(T29, S0)
    fd.add_output(T30)


with FusionDefinition() as fd:
    nvfuser_fusion_id9(fd)

for _ in range(5):
    out = fd.execute(inputs)
```

### PR DESCRIPTION
**Pull request by wujingyue**
This is a refactoring PR needed to fix #775 . A separate PR will change FusionExecutor to return both PTX and SASS.

### CODE REVIEW

**Comment by wujingyue** [comment]
Apparently, clang-tidy failed for an unrelated reason. 

```
  Could not find a package configuration file provided by "Torch" with any of
  the following names:

    TorchConfig.cmake
    torch-config.cmake
```

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the cleanup

**Comment by wujingyue** [comment]
> Apparently, clang-tidy failed for an unrelated reason. cc @xwang233
> 
> ```
>   Could not find a package configuration file provided by "Torch" with any of
>   the following names:
> 
>     TorchConfig.cmake
>     torch-config.cmake
> ```

@xwang233 I reran the clang-tidy check 10 minutes ago and it seems to work as expected -- shows me some real errors.

**Comment by wujingyue** [comment]
!build

**Comment by xwang233** [comment]
The error in clang-tidy cmake yesterday could be issue from Meta's torch wheel, which might be fixed today? If your rerun works, it's fine.

**Comment by wujingyue** [comment]
> The error in clang-tidy cmake yesterday could be issue from Meta's torch wheel, which might be fixed today? If your rerun works, it's fine.

Yeah. I figured it might be a transient error. Thanks for checking!
